glusterblock/glusterblock-provisioner.go

```diff
@@ -0,0 +1,355 @@
+		case "resturl":
+			cfg.url = v
+		case "restuser":
+			cfg.user = v
+		case "restuserkey":
+			cfg.userKey = v
+		case "secretname":
+			cfg.secretName = v
+		case "secretnamespace":
+			cfg.secretNamespace = v
+		case "clusterid":
+			if len(v) != 0 {
+				cfg.clusterId = v
+			}
+		case "restauthenabled":
+			authEnabled = dstrings.ToLower(v) == "true"
+		case "hacount":
+			haCount, err = strconv.Atoi(v)
+			if err != nil {
+				return nil, fmt.Errorf("glusterblock: failed to parse hacount %v ", k)
+			}
+		case "opmode":
+			cfg.opMode = v
+		case "execpath":
+			cfg.execPath = v
+		default:
+			return nil, fmt.Errorf("glusterblock: invalid option %q for volume plugin %s", k, "glusterblock")
+		}
+	}
+
+	if len(cfg.url) == 0 {
+		return nil, fmt.Errorf("StorageClass for provisioner %s must contain 'resturl' parameter", "glusterblock")
+	}
+
+	if haCount == 0 {
+		cfg.haCount = haCount
+	}
+
+	if len(cfg.opMode) == 0 {
+		cfg.opMode = "executable"
+	}
+
+	if len(cfg.execPath) == 0 {
+		cfg.execPath = defaultExecPath
+	}
+	if !authEnabled {
+		cfg.user = ""
+		cfg.secretName = ""
+		cfg.secretNamespace = ""
+		cfg.userKey = ""
+		cfg.secretValue = ""
+	}
+
+	if len(cfg.secretName) != 0 || len(cfg.secretNamespace) != 0 {
+		// secretName + Namespace has precedence over userKey
+		if len(cfg.secretName) != 0 && len(cfg.secretNamespace) != 0 {
+			cfg.secretValue, err = parseSecret(cfg.secretNamespace, cfg.secretName, kubeclient)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, fmt.Errorf("StorageClass for provisioner %q must have secretNamespace and secretName either both set or both empty", "glusterblock")
+		}
+	} else {
+		cfg.secretValue = cfg.userKey
+	}
+
+	return &cfg, nil
+}
+
+
+// parseSecret finds a given Secret instance and reads user password from it.
+func parseSecret(namespace, secretName string, kubeClient kubernetes.Interface) (string, error) {
+
+	secretMap, err := GetSecretForPV(namespace, secretName, "glusterblock", kubeClient)
+	if err != nil {
+		glog.Errorf("failed to get secret %s/%s: %v", namespace, secretName, err)
+		return "", fmt.Errorf("failed to get secret %s/%s: %v", namespace, secretName, err)
+	}
+	if len(secretMap) == 0 {
+		return "", fmt.Errorf("empty secret map")
+	}
+	secret := ""
+	for k, v := range secretMap {
+		if k == secretKeyName {
+			return v, nil
+		}
+		secret = v
+	}
+	// If not found, the last secret in the map wins as done before
+	return secret, nil
+}
+
+
+// GetSecretForPV locates secret by name and namespace, verifies the secret type, and returns secret map
+func GetSecretForPV(secretNamespace, secretName, volumePluginName string, kubeClient clientset.Interface) (map[string]string, error) {
+        secret := make(map[string]string)
+        if kubeClient == nil {
+                return secret, fmt.Errorf("Cannot get kube client")
+        }
+        secrets, err := kubeClient.Core().Secrets(secretNamespace).Get(secretName, metav1.GetOptions{})
```

Comment:
it says metav1 doesn't exist, probably it's supposed to be an alias for "k8s.io/apimachinery/pkg/apis/meta/v1" like in kubernetes code?